### PR TITLE
Add-cyw43-(blinky-example)

### PR DESCRIPTION
**Pull request by SMartinScottLogic**

### CODE REVIEW

**Comment by SMartinScottLogic** [comment]
Do we want a point-in-time firmware added to the repo, to allow automatic check (linting) to run?

File: src/fat12_partition/setup.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+static FS_DUMP: &[u8; 102400] = include_bytes!("../../dumps/linux_partitioned.dump");
+pub fn init(storage: &mut [u8], block_size: u32, blocks: u32) {
+    assert!(storage.len() >= (blocks * block_size) as usize);
+    assert!(storage.len() >= FS_DUMP.len());
+    unsafe {
+        let dst = storage.as_mut_ptr();
+        core::ptr::copy(FS_DUMP.as_ptr(), dst, FS_DUMP.len());
+    }
```
Comment (bobrippling) [inline src/fat12_partition/setup.rs]: Here we could avoid `unsafe` with a bit of:

```suggestion
    storage[..FS_DUMP.len()].copy_from_slice(FS_DUMP);
```


(and the two in [`fat12_partition/mod.rs`](https://github.com/chrisprice/pico-usb-mass-storage/blob/b6b49925bf0e442cf893092678527822a84483b2/src/fat12_partition/mod.rs#L174-L195))

File: src/main.rs [hunk 2]
```diff
@@ -103,7 +103,7 @@ async fn main(spawner: Spawner) {
     let blinky_fut = blinky.run();
     // Run everything concurrently.
     // If we had made everything `'static` above instead, we could do this using separate tasks instead.
-    join(join(usb_fut, usb_mass_storage_fut), blinky_fut).await;
+    join_array([usb_fut, usb_mass_storage_fut, blinky_fut]).await;
```
Comment (SMartinScottLogic) [inline src/main.rs]: Needs reverting - Futures of inconsistent types :( (Future<Output = !> vs Future <Output = ()> )

**Comment by bobrippling** [comment]
> Do we want a point-in-time firmware added to the repo, to allow automatic check (linting) to run?

Yeah I'd say so - that's what embassy do too, in fact

File: src/main.rs (hunk 2 above)
Comment (bobrippling) [reply src/main.rs]: We can make a few little changes to sort this - I can suggest here or open a PR to your repo/this branch, any preference?

File: src/main.rs [hunk 3]
```diff
@@ -47,10 +42,25 @@ impl State {
 }
 
 #[embassy_executor::main]
-async fn main(_spawner: Spawner) {
-    let p = embassy_rp::init(Default::default());
+async fn main(spawner: Spawner) {
+    unsafe {
+        fat12_partition::init(&mut STORAGE, BLOCK_SIZE, BLOCKS);
+    }
```
Comment (bobrippling) [inline src/main.rs]: Perhaps we isolate the `unsafe` to show just the bit we're unsafeing:
```suggestion
    fat12_partition::init(unsafe { &mut STORAGE }, BLOCK_SIZE, BLOCKS);
```

(and at the end of this file too)

File: src/fat12_partition/mod.rs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    fn read1(&mut self) -> u8 {
+        let value = self.data[self.position as usize];
+        self.position += 1;
+        value
+    }
+    fn read4(&mut self) -> u32 {
+        let value = ((self.data[(self.position + 3) as usize] as u32) << 24)
+            + ((self.data[(self.position + 2) as usize] as u32) << 16)
+            + ((self.data[(self.position + 1) as usize] as u32) << 8)
+            + (self.data[self.position as usize] as u32);
```
Comment (bobrippling) [inline src/fat12_partition/mod.rs]: Here we could use [`from_le_bytes`](https://doc.rust-lang.org/std/primitive.u32.html#method.from_le_bytes) on a slice, something like:

```rust
let position = self.position as usize;
let slice = self.data[position .. position + 4];
let value = u32::from_le_bytes(slice.try_into().unwrap());
```

File: src/main.rs (hunk 2 above)
Comment (SMartinScottLogic) [reply src/main.rs]: I've reverted for now. Happy for you to suggest here.

Comment (bobrippling) [reply src/main.rs]: Sweet, for the `()` vs `!`, we can propagate the never out from bbb so they're all `Future<Output = !>`:

```diff
diff --git a/src/scsi/mod.rs b/src/scsi/mod.rs
@@ -85,9 +85,9 @@ impl<'d, D: Driver<'d>, M: RawMutex> Scsi<'d, D, M> {
         }
     }
 
-    pub async fn run(&mut self, handler: &mut impl Handler) {
+    pub async fn run(&mut self, handler: &mut impl Handler) -> ! {
         let mut adapter = Adapter { handler };
-        self.transport.run(&mut adapter).await;
+        self.transport.run(&mut adapter).await
     }
 }
 
diff --git a/src/usb_mass_storage/mod.rs b/src/usb_mass_storage/mod.rs
@@ -81,7 +81,7 @@ impl<'d, D: Driver<'d>, M: RawMutex> UsbMassStorage<'d, D, M> {
         }
     }
 
-    pub async fn run(&mut self, handler: &mut impl Handler) {
+    pub async fn run(&mut self, handler: &mut impl Handler) -> ! {
         self.scsi.run(handler).await
     }
 }
```


---

Also If you like, for this bit we can use `join3`:

```rust
join3(usb_fut, usb_mass_storage_fut, blinky_fut).await;
```

**Comment by bobrippling** [comment]
Merged in 3b1f1dea91de2685099c3d256271a5d83fedca71

**Pull request closed without merging**